core/arch/arm/emit_utils.c

```diff
@@ -980,10 +980,26 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
     add_patch_marker(patch, target_delete_entry, PATCH_ASSEMBLE_ABSOLUTE,
                      0 /* beginning of instruction */,
                      (ptr_uint_t *)&ibl_code->target_delete_entry);
+
     /* We just executed the hit path, so the app's r1 and r2 values are still in
      * their TLS slots, and &linkstub is still in the r3 slot.
      */
 
+    /* i#4665: We speculatively store next fragment tag in r2 (stored in dcontext later)
+     * and &linkstub_ibl_deleted in r1 (passed to fcache_return via r0 later). This can be
+     * tested on ibl-stress test. But that hasn't been ported to ARM yet, and we do not
+     * have an ARM machine available.
+     */
+    ASSERT_NOT_TESTED();
+    APP(&ilist,
+        INSTR_CREATE_ldr(
+            dc, opnd_create_reg(DR_REG_R2),
+            OPND_CREATE_MEMPTR(DR_REG_R1, offsetof(fragment_entry_t, tag_fragment))));
+    instrlist_insert_mov_immed_ptrsz(dc, (ptr_uint_t)get_ibl_deleted_linkstub(),
+                                     opnd_create_reg(DR_REG_R1), &ilist, NULL, NULL,
+                                     NULL);
+    APP(&ilist, instr_create_restore_from_tls(dc, DR_REG_R0, TLS_REG0_SLOT));
```

Comment:
At this point, `r0` should contain the `target_delete_entry` address, set by the IBL hit path. L1009 would overrwrite app's `r0` value, unless this new line at L1001 is added.

Reply:
Yes, this makes sense: the shared unlink: sequence will save r0.

Reply:
Should I take this branch and try to run a sanity check on my armbook?

Reply:
Is there a test you suggest running?  ibl-stress is not ported?

Reply:
`ibl-stress` is one option, but it isn't ported yet to ARM. Another option is to set `-shared_ibt_table_bb_init` to a very low value, and try with whatever large app is available.

Reply:
Testing on some ARM machine would be great. Thanks for the offer. I was a bit uneasy putting in  many changes in ARM without testing them.

Reply:
Let me quickly try and check if any test suite app can trigger this path with a low shared_ibt_table_bb_init.

Reply:
You can use these. I think they are available on ARM.
```
dynamorio/bin64/drrun "-s" "90" "-quiet" "-debug" "-killpg"  "-dumpcore_mask" "0" "-enable_reset" "-reset_at_nth_thread" "2" "-code_api" "-disable_traces" "-shared_bb_ibt_tables" "-shared_ibt_table_bb_init" "2" "--" "dynamorio/suite/tests/bin/linux.clone"
```

```
dynamorio/bin64/drrun "-s" "90" "-quiet" "-debug" "-killpg"  "-dumpcore_mask" "0" "-code_api" "-disable_traces" "-shared_bb_ibt_tables" "-shared_ibt_table_bb_init" "2" "--" "dynamorio/suite/tests/bin/linux.signest"
```

Reply:
These are the `linux.clone-reset` and `linux.signest` tests respectively, with their options modified to include `-disable_traces -shared_bb_ibt_tables -shared_ibt_table_bb_init 2` in `vmap_run_list` [SHORT](https://github.com/DynamoRIO/dynamorio/blob/a9be2914cd1971f7d4292101a21ff3b18fcf877b/suite/tests/CMakeLists.txt#L101) configuration.

Reply:
The first crashes with reset enabled, but that is listed as a known ARM problem.  Without reset it runs and it does hit the ASSERT_NOT_TESTED (though I guess that's on generation so any app at init will hit it):

```
> bin32/drrun -disable_traces -loglevel 1 "-shared_bb_ibt_tables" "-shared_ibt_table_bb_init" "2" "--" "suite/tests/bin/linux.clone"
<log dir=/home/derek/dr/git/build_dbg_tests/bin32/../logs/linux.clone.28638.00000000>
<Starting application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.clone (28638)>
<Not tested @/home/derek/dr/git/src/core/arch/arm/emit_utils.c:993>
<Initial options = -no_dynamic_options -loglevel 1 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -shared_bb_ibt_tables -shared_ibt_table_bb_init 2 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<(1+x) Handling our fault in a TRY at 0x711e60e0>
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/home/derek/dr/git/build_dbg_tests/lib32/debug/libdynamorio.so' 0x71024368
>
Sideline thread started
i = 2500000
i = 5000000
i = 7500000
i = 10000000
i = 12500000
i = 15000000
i = 17500000
i = 20000000
i = 22500000
i = 25000000
Sideline thread finished
Waiting for child to exit
Child has exited
Sideline thread started
i = 2500000
i = 5000000
i = 7500000
i = 10000000
i = 12500000
i = 15000000
i = 17500000
i = 20000000
i = 22500000
i = 25000000
Sideline thread finished
Waiting for child to exit
Child has exited
<Stopping application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.clone (28638)>
```

It has no exit hits but it's probably too small with too few threads:
```
$ egrep 'table resiz|deleted but hit' `ls -1td logs/*0|head -1`/li*
```

Reply:
The other hits a problem that does not occur without the ibt table options so may indicate a problem with this PR?

```
> bin32/drrun -disable_traces -loglevel 1 "-shared_bb_ibt_tables" "-shared_ibt_table_bb_init" "2" "--" "suite/tests/bin/linux.signest"
<log dir=/home/derek/dr/git/build_dbg_tests/bin32/../logs/linux.signest.29007.00000000>
<Starting application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.signest (29007)>
<Not tested @/home/derek/dr/git/src/core/arch/arm/emit_utils.c:993>
<Initial options = -no_dynamic_options -loglevel 1 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -shared_bb_ibt_tables -shared_ibt_table_bb_init 2 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<(1+x) Handling our fault in a TRY at 0x711e60e0>
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/home/derek/dr/git/build_dbg_tests/lib32/debug/libdynamorio.so' 0x71024368
>
sending 2 signals
in handler 42
in handler 42
<get_memory_info mismatch! (can happen if os combines entries in /proc/pid/maps)
	os says: 0xb6475000-0xb6675000 prot=0x00000000
	cache says: 0xb6475000-0xb6676000 prot=0x00000000
>
<Application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.signest (29007).  Internal Error: DynamoRIO debug check failure: /home/derek/dr/git/src/core/unix/signal.c:5452 syscall_signal || safe_is_in_fcache(dcontext, pc, (byte *)sc->SC_XSP)
(Error occurred @2526 frags)
version 8.0.18659, custom build
-no_dynamic_options -loglevel 1 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -shared_bb_ibt_tables -shared_ibt_table_bb_init 2 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct 
/home/derek/dr/git/build_dbg_tests/lib32/debug/libdynamorio.so=0x71000000>
```

Reply:
Trying to analyze this crash...the gdb on this machine is old and having symbol trouble.  The big-little core is also missing some kernel support like watchpoints IIRC.

Reply:
About `linux.clone`, without reset it doesn't hit the bug, so a successful run there doesn't tell us anything. `Fcache exits, deleted but hit in ibl` should also be zero for this run.

Reply:
About the `linux.signest` crash, does it also occur with a high value of `-shared_ibt_table_bb_init`? (which shouldn't invoke the new code, but still use shared ibt)

Reply:
linux.signest does not crash with the default value for `-shared_ibt_table_bb_init`.  In gdb: it does seem to hit the target_delete case and goes to that IBL code sequence.  I think the bug is that our gencode is Thumb but it does a `bx` without LSB set.

Reply:
OK after fixing the Thumb LSB issue we hit a crash that is a bug in this PR:

```
Thread 1 "linux.signest" received signal SIGSEGV, Segmentation fault.
0x41e1d35e in ?? ()
1: x/i $pc
=> 0x41e1d35e:	ldr	r2, [r1, #0]
```

r1 was cleared on the hit path up above:
```
1: x/i $pc
=> 0x41e1d334:	sub.w	r0, r0, r2
(gdb) 
0x41e1d338 in ?? ()
1: x/i $pc
=> 0x41e1d338:	cbnz	r0, 0x41e1d346
(gdb) 
0x41e1d33a in ?? ()
1: x/i $pc
=> 0x41e1d33a:	ldr	r0, [r1, #4]
(gdb) 
0x41e1d33c in ?? ()
1: x/i $pc
=> 0x41e1d33c:	ldr.w	r1, [r10, #4]
(gdb) 
0x41e1d340 in ?? ()
1: x/i $pc
=> 0x41e1d340:	ldr.w	r2, [r10, #8]
(gdb) 
0x41e1d344 in ?? ()
1: x/i $pc
=> 0x41e1d344:	bx	r0
```

Hmm.  Looking at x86: it doesn't load the target into rcx; it jumps through memory `jmp 0x08(%rcx)`, so rcx still holds the table pointer.  So it looks like I was wrong: we don't have the table pointer still available after the hit jump for aarchxx??

Reply:
So it is there for a64, so that does hold; it's just a32 that does not have it.

Reply:
So for a32 we would either:

1) Re-arrange the IBL to have the table pointer in r0 (swapping with r1 for the table load and compare) and change the `bx r0` to `ldr pc, [r0 + 4]`
2) Add a store on the hit path to stick the tag r2 into a TLS slot to be restored on the target_delete path

?

Reply:
Ah right. ARM restores app x1 at L936. Sorry, I should've seen that. Let me put in a proper workaround.

Reply:
So, in this case, I guess we don't have an option but to store the tag into a TLS slot and take the performance hit on the hit path right?

Reply:
> So, in this case, I guess we don't have an option but to store the tag into a TLS slot and take the performance hit on the hit path right?

See https://github.com/DynamoRIO/dynamorio/pull/4699#discussion_r568238106 which gives another option which I assume is more performant (that cache line is already in L1 and the load should beat a store) but gets more complex w/ rearranging the IBL.

Reply:
I pushed my Thumb LSB fix to this branch.  I guess add a `Co-authored-by: ` to the final squash message.

Reply:
Thanks Derek for the help in testing this on ARM and the fix too.

I've pushed my fix to this branch; can you please test again? I have implemented option 2 above for now, as it is simpler, given that I don't have access to an ARM machine.

Reply:
It gets through two target_deletes (shouldn't it have the new table after going back to DR?  not sure I understand it) and then hits an assert on the mode.  Not sure what the cause is: the IBL has insert_mode_change_handling.
```
Entry into F2526(0x00010e18).0x4cb9de90 (A32)(shared)
Exit from fragment deleted but hit in ibl

d_r_dispatch: target = 0x00010d74
Entry into F1383(0x00010d74).0x4caf6f10 (A32)(shared)
Exit from fragment deleted but hit in ibl

d_r_dispatch: target = 0xb6fb90c8
Entry into F1240(0xb6fb90c8).0x4caf42bc (T32)(shared)
SYSLOG_ERROR: Application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.signest (28957).  Internal Error: DynamoRIO debug check failure: /home/derek/dr/git/src/core/dispatch.c:487 dr_get_isa_mode(dcontext) == FRAG_ISA_MODE(targetf->flags) IF_X64(|| (dr_get_isa_mode(dcontext) == DR_ISA_IA32 && !FRAG_IS_32(targetf->flags) && DYNAMO_OPTION(x86_to_x64)))
(Error occurred @2526 frags)
```

Reply:
> It gets through two target_deletes (shouldn't it have the new table after going back to DR? not sure I understand it)
You mean the same thread invokes `target_delete_entry` code multiple times? That could happen if the IBL goes through multiple resizes I think.

Reply:
See the log snippet I pasted at https://github.com/DynamoRIO/dynamorio/pull/4699#discussion_r568296180: there is no resize in between them (a resize prints a line at this loglevel: I see them up above in the log).  There's nothing in between them but a roundtrip from dispatch to the cache.

Reply:
Ah, I see the source of the assert: insert_mode_change_handling() is called very early in the IBL and it clears the LSB.  Then later in the code after the unlinked: entry point it's called again, but now using the LSB-cleared copy of the tag.  So that needs to be fixed.  Re-loading from the table would fix it since the table has the LSB-set version.  Not sure how else to easily fix it: an inverted form of insert_mode_change_handling(), or storing the original in yet another TLS slot, or something.

Reply:
If the issue is a duplicate invocation of `insert_mode_change_handling`, then c42836e may be a possible fix that I pushed to the branch now.

Reply:
Yes that works:

```
$ bin32/drrun -disable_traces -loglevel 2 "-shared_bb_ibt_tables" "-shared_ibt_table_bb_init" "2" "--" "suite/tests/bin/linux.signest"
<log dir=/home/derek/dr/git/build_dbg_tests/bin32/../logs/linux.signest.4232.00000000>
<Starting application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.signest (4232)>
<Initial options = -no_dynamic_options -loglevel 2 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -shared_bb_ibt_tables -shared_ibt_table_bb_init 2 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<(1+x) Handling our fault in a TRY at 0x711e6070>
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/home/derek/dr/git/build_dbg_tests/lib32/debug/libdynamorio.so' 0x71024368
>
sending 2 signals
in handler 42
in handler 42
<get_memory_info mismatch! (can happen if os combines entries in /proc/pid/maps)
	os says: 0xb6458000-0xb6658000 prot=0x00000000
	cache says: 0xb6458000-0xb6659000 prot=0x00000000
>
all done
<Stopping application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/linux.signest (4232)>

$ egrep -n 'deleted but hit|bb hashtable resiz' `ls -1td logs/*0|head -1`/lo*
logs/linux.signest.4232.00000000/log.0.4232.html:2819:indjmp_bb hashtable resized at 4 entries from capacity 5 to 9
logs/linux.signest.4232.00000000/log.0.4232.html:3293:ret_bb hashtable resized at 4 entries from capacity 5 to 9
logs/linux.signest.4232.00000000/log.0.4232.html:3638:indjmp_bb hashtable resized at 7 entries from capacity 9 to 17
logs/linux.signest.4232.00000000/log.0.4232.html:4738:ret_bb hashtable resized at 7 entries from capacity 9 to 17
logs/linux.signest.4232.00000000/log.0.4232.html:6598:shared_bb hashtable resized at 564 entries from capacity 1025 to 2049
logs/linux.signest.4232.00000000/log.0.4232.html:7016:ret_bb hashtable resized at 12 entries from capacity 17 to 33
logs/linux.signest.4232.00000000/log.0.4232.html:8330:indcall_bb hashtable resized at 4 entries from capacity 5 to 9
logs/linux.signest.4232.00000000/log.0.4232.html:8703:indjmp_bb hashtable resized at 12 entries from capacity 17 to 33
logs/linux.signest.4232.00000000/log.0.4232.html:9249:ret_bb hashtable resized at 24 entries from capacity 33 to 65
logs/linux.signest.4232.00000000/log.0.4232.html:9831:ret_bb hashtable resized at 46 entries from capacity 65 to 129
logs/linux.signest.4232.00000000/log.0.4232.html:10371:indcall_bb hashtable resized at 7 entries from capacity 9 to 17
logs/linux.signest.4232.00000000/log.0.4232.html:13523:indjmp_bb hashtable resized at 24 entries from capacity 33 to 65
logs/linux.signest.4232.00000000/log.0.4232.html:13934:shared_bb hashtable resized at 1127 entries from capacity 2049 to 4097
logs/linux.signest.4232.00000000/log.0.4232.html:15035:ret_bb hashtable resized at 91 entries from capacity 129 to 257
logs/linux.signest.4232.00000000/log.0.4232.html:18075:indcall_bb hashtable resized at 12 entries from capacity 17 to 33
logs/linux.signest.4232.00000000/log.0.4232.html:20911:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:20915:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:20919:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:20923:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:20989:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:20993:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:20997:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21001:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21034:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21038:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21049:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21053:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21057:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21061:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21065:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21079:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21253:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21540:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21573:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21584:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21595:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21793:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21797:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21974:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:21978:Exit from fragment deleted but hit in ibl
logs/linux.signest.4232.00000000/log.0.4232.html:22736:              Fcache exits, deleted but hit in ibl (thread):       25
logs/linux.signest.4232.00000000/log.1.4234.html:2008:indcall_bb hashtable resized at 24 entries from capacity 33 to 65
logs/linux.signest.4232.00000000/log.1.4234.html:4518:shared_bb hashtable resized at 2254 entries from capacity 4097 to 8193
logs/linux.signest.4232.00000000/log.1.4234.html:5513:indjmp_bb hashtable resized at 46 entries from capacity 65 to 129

$ egrep -n 'deleted but hit|resiz' `ls -1td logs/*0|head -1`/li* | tail -3
3771:                                       IBT resizes :       15
3781:          Extra exits due to shared table resizing :        1
3788:              Fcache exits, deleted but hit in ibl :       25
```

Reply:
I still don't understand why there are so many "Exit from fragment deleted but hit in ibl" when there are no further resizes in between -- does that make sense to you?

Reply:
The other thread did perform 3 resizes: but we have a lot of exits in a row, far more than that would explain.

Reply:
I see those `hit-but-exit`s in a row on AArch64 too.

A possibility is that there's some bug that's causing misses all the time after the resize; maybe some missing update of table/mask? A run of `ibl-stress` with 20k sequences and `-shared_ibt_table_bb_init 2` gives 455 `Exit from fragment deleted but hit in ibl` logs. Given this (perhaps low) number, is such a bug likely? `ibl-stress` has 1 jump reg and 1 call reg per sequence. The exits due to resizing is quite high though.

```
./api.ibl-stress.0.29838.html:          Extra exits due to shared table resizing :             58001
```
```
./api.ibl-stress.0.29838.html:              Fcache exits, deleted but hit in ibl :               455

```

Or maybe the resize took so long that this thread executed `target_delete_entry` multiple times? Given that the thread would need to go through dispatch each time (which would perhaps give more time for the resize to complete), is that likely? 

Logs from ibl-stress run with high sequence count: Notice the bursty-ness, which makes me think maybe the resize did take too long.
```
./log.3.29847.html:1485:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1489:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1493:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1497:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1501:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1505:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1509:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1513:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1517:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1521:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1525:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1529:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1533:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1537:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1549:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1553:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1557:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1561:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1565:Exit from fragment deleted but hit in ibl
./log.3.29847.html:1630:Exit from fragment deleted but hit in ibl

./log.3.29847.html:19138:Exit from fragment deleted but hit in ibl
./log.3.29847.html:19142:Exit from fragment deleted but hit in ibl
./log.3.29847.html:19146:Exit from fragment deleted but hit in ibl

./log.3.29847.html:38570:Exit from fragment deleted but hit in ibl
./log.3.29847.html:38574:Exit from fragment deleted but hit in ibl
./log.3.29847.html:38578:Exit from fragment deleted but hit in ibl
./log.3.29847.html:38582:Exit from fragment deleted but hit in ibl
./log.3.29847.html:38586:Exit from fragment deleted but hit in ibl

./log.3.29847.html:77637:Exit from fragment deleted but hit in ibl
./log.3.29847.html:77641:Exit from fragment deleted but hit in ibl
./log.3.29847.html:77645:Exit from fragment deleted but hit in ibl
```

Reply:
> Or maybe the resize took so long that this thread executed target_delete_entry multiple times? Given that the thread would need to go through dispatch each time (which would perhaps give more time for the resize to complete), is that likely?

I added an `os_thread_sleep(1000);` to dispatch when it hits the `Exit from fragment deleted but hit in ibl` to give the other thread time to finish a resize: yet I have the same # of such exits in a row.

Reply:
> Are you aware of any dispatch flows where dcontext->last_exit is not updated? That could cause duplicate counting and logs.

To check if this is the cause, I set `dcontext->last_exit` to a new linkstub after the log is printed in `dispatch_exit_fcache_stats`, but I still see the exits in a row.

Reply:
Perhaps this has something to do with the weak memory model on arm? But then your  `os_thread_sleep` experiment should have shown some effect on the # of repeated logs.

Reply:
https://github.com/DynamoRIO/dynamorio/blob/a9be2914cd1971f7d4292101a21ff3b18fcf877b/core/fragment.c#L4217

Here it's documented that a thread's local ptr and state are updated only on IBL miss. Not sure if that includes ibl hits with `target_delete` payload too or not. If it doesn't, that may explain the extra "ibl hits but exit" after resize.

Reply:
Looks like the special linkstub is not marked indirect, so it doesn't go into that code.  But fixing it results in an assert somewhere else.  Let's separate it out from this PR.  I filed #4704.  I think this PR is good to go.

Reply:
Okay, I'll go ahead with this PR then. Thanks for flagging #4704.